llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp

```diff
@@ -1398,8 +1398,8 @@ class AMDGPULowerModuleLDS {
           LDSVarsToTransform.begin(), LDSVarsToTransform.end()));
 
       for (GlobalVariable *GV : Sorted) {
-        OptimizedStructLayoutField F(GV,
-                                     DL.getTypeAllocSize(GV->getValueType()),
+        unsigned Size = DL.getTypeAllocSize(GV->getValueType());
+        OptimizedStructLayoutField F(GV, Size ? Size : 1,
```

Comment:
What is the meaning of a 0 size allocation? Can we just not add it to LayoutFields instead of wasting a byte?

Reply:
It's really an edge case, it can happen if the compiler allows something like `static unsigned arr[0]` (zero-sized externals become dynlds I think)
I can try to not add it and refactor the code as needed to make it work, I just thought it wasn't really worth it for such an edge case.

Reply:
Yes, it's an edge case but what does it actually mean?

Reply:
I think it'd be for a variable length array. It's why we lower as dynlds for external ones. I assume we just can't lower them then and we should probably just fatal_error. It's a bit ugly to fatal in the backend but better than asserting

Reply:
I think the correct solution is to not add it to the layout, and just track there was this unknown size object. All the 0 sized objects are supposed to receive the same address (which we would want at the end of the known allocation)